Add getting started guide for App Platform (#912)
giantswarm/docs: src/content/app-platform/getting-started/index.md

---
linkTitle: Getting started
title: Getting started with App Platform
description: Guide to deploying apps using kubectl and the Giant Swarm Management API.
weight: 20
menu:
  main:
    parent: app-platform
last_review_date: 2021-04-21
owner:
  - https://github.com/orgs/giantswarm/teams/team-batman
user_questions:
  - How can I deploy an app using an App CR?
  - How can I configure an App CR?
  - How can I see which apps are available using the Management API?
---

# Getting started with App Platform

## Overview

> 

Reviewer: ```suggestion
App Platform allows you to manage apps and their configurations (represented by App Custom Resources or App CRs) in multiple clusters, from a single place: the management cluster.
```

Would it be useful to add a one-line reminder of what app platform is?

Author: Thanks @cokiengchiara yes the intro needed more context.

I changed it to include Helm. WDYT? 

> _Giant Swarm App Platform_ is built on top of Helm

If this wording is OK I think we should also use it in https://docs.giantswarm.io/app-platform/overview/

Reviewer: yes sounds perfect. Great catch on modifying also in that doc. Could you include in this PR? 😅

> In this guide we will install an ingress controller which helps you expose your

Reviewer: In this guide, we will install the NGINX ingress controller app. We'll do this by using kubectl, which creates App CRs using the Kubernetes API of your management cluster.

App CRs can be created this way via your automation or our Web UI.

In general, you can manage App CRs with any tool that can communicate with the Kubernetes API such as Helm or GitOps tools (like Argo CD or Flux).

Reviewer: I'm suggesting this flow that for me flows more easily: 

"Here, we use kubectl which uses kubernetes api. this can also be done via automation and web ui. in fact, it can be done using any tool that can communicate with the k8s api!"

Author: Yes, that flows better.

services to the outside world.

We already have a [guide]({{< relref "/getting-started/ingress-controller" >}})
for doing this using our web UI. For this guide we will use kubectl to create
[App]({{< relref "/ui-api/management-api/crd/apps.application.giantswarm.io.md" >}})
custom resources (CRs) using the Kubernetes API of your
[management cluster]({{< relref "/general/management-clusters" >}}).

However you can manage the App CRs with any tool that can communicate
with the Kubernetes API such as Helm or GitOps tools (like Argo CD or Flux).

## Setting up

You can access your management cluster using the [kubectl gs login]({{< relref "/ui-api/kubectl-gs/login" >}})
command of our kubectl plugin. See [here]({{< relref "/ui-api/kubectl-gs/installation" >}})
for how to install it.

In the management cluster your App CRs are stored in a namespace with the same
name as your workload cluster ID. Let's set an environment variable for this
which we will use in the later steps.

```sh

> export NAMESPACE=** your cluster id**

Reviewer: ```suggestion
export NAMESPACE=**your cluster id**
```

```sh
kubectl get appcatalogentry -l latest=true,application.giantswarm.io/catalog=giantswarm

NAME                                             CATALOG      APP NAME                       APP VERSION   VERSION        AGE
...
giantswarm-nginx-ingress-controller-app-1.16.1   giantswarm   nginx-ingress-controller-app   v0.45.0       1.16.1         25h
...
```

## Creating an App CR

We can use the [kubectl gs template app]({{< relref "/ui-api/kubectl-gs/login" >}})
command to generate the App CR using the latest version from the previous command.

```sh
kubectl gs template app \
  --catalog=giantswarm \
  --cluster=${NAMESPACE} \
  --name=nginx-ingress-controller-app \
  --namespace=kube-system \
  --version=1.16.1 > nginx-ingress-controller-app.yaml

kubectl apply -f nginx-ingress-controller-app.yaml
```

Lets first see the output of the template command which shows only the
required fields.

```yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  name: nginx-ingress-controller-app
  namespace: s2wf9
spec:
  catalog: giantswarm
  kubeConfig:
    inCluster: false
  name: nginx-ingress-controller-app
  namespace: kube-system
  version: 1.16.1

> ```

Author: I'd like `template app` to check if app-admission-controller is enabled for the cluster. If so it only defaults the required fields. As the other details make it harder to understand the app CRs.

## Defaulting and App Status

Now lets check the app using the `kubectl gs get app` command which filters
the output to show only the most important fields.

```sh

> kubectl -n ${NAMESPACE} gs get app nginx-ingress-controller-app -o yaml

Author: I think this will be useful for customers but mainly I want it for myself 😅 

I use the `kubectl neat` plugin to get rid of the managed fields and other stuff I don't care about but it also removes the CR status which is a pain.

https://github.com/itaysk/kubectl-neat

```

The labels, cluster config and kubeconfig have been all defaulted to the correct
values for your cluster. You can read more about defaulting and validation for
App CRs [here]({{< relref "/app-platform/defaulting-validation" >}}).

```yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  labels:
    app-operator.giantswarm.io/version: 3.2.1
    app.kubernetes.io/name: nginx-ingress-controller-app
  name: nginx-ingress-controller-app
  namespace: s2wf9
spec:
  catalog: giantswarm
  config:
    configMap:
      name: ingress-controller-values
      namespace: s2wf9
  kubeConfig:
    context:
      name: s2wf9
    inCluster: false
    secret:
      name: s2wf9-kubeconfig
      namespace: s2wf9
  name: nginx-ingress-controller-app
  namespace: kube-system
  version: 1.16.1
status:
  appVersion: v0.45.0
  release:
    lastDeployed: "2021-04-21T16:28:08Z"
    status: deployed
  version: 1.16.1
```

In the App CR status you can see that the app is deployed. The `appVersion`
shows that this version of the app is deploying `v0.45.0` of the upstream

> [Nginx Ingress Controller](https://github.com/kubernetes/ingress-nginx) project.

Author: @cokiengchiara I added this so we can close https://github.com/giantswarm/giantswarm/issues/11992

Reviewer: perfect thank you!

Reviewer: thinking if we should state what `version` is but maybe that's obvious.

Author: I think version is OK. The problem is appVersion because lots of things are called app 😂

---
linkTitle: Getting started
title: Getting started with App Platform
description: Guide to deploying apps using kubectl gs and the Giant Swarm Management API.
weight: 20
menu:
  main:
    parent: app-platform
last_review_date: 2021-07-07
owner:
  - https://github.com/orgs/giantswarm/teams/team-batman
user_questions:
  - How can I deploy an app using an App CR?
  - How can I configure an App CR?
  - How can I see which apps are available using the Management API?
---

# Getting started with App Platform

## Overview

> The _Giant Swarm App Platform_ is built on top of [Helm](https://helm.sh/) and allows you to manage apps and their configurations represented by App Custom Resources (CRs) for multiple clusters, from a single place: the [Management API]({{< relref "/management-api/overview" >}}).

Reviewer 2: ```suggestion
The _Giant Swarm App Platform_ is built on top of [Helm](https://helm.sh/) and allows you to manage apps and their configurations represented by App Custom Resources (CRs) for multiple clusters, from a single place: the [Management API]({{< relref "/ui-api/management-api/overview" >}}).
```

In this guide, we will install the NGINX Ingress Controller app. We will do this by using kubectl, to create
an [App]({{< relref "/ui-api/management-api/crd/apps.application.giantswarm.io.md" >}}) CR using the Kubernetes API of your management cluster.

App CRs can be created this way via your automation or our Web UI (See: [guide]({{< relref "/getting-started/ingress-controller" >}})).

In general, you can manage App CRs with any tool that can communicate with the Kubernetes API such as Helm or GitOps tools (like Argo CD or Flux CD).

## Setting up

You can access your management cluster using the [kubectl gs login]({{< relref "/ui-api/kubectl-gs/login" >}})
command of our kubectl plugin. See [here]({{< relref "/ui-api/kubectl-gs/installation" >}})
for how to install it. If you already have it installed please upgrade to the
latest version.

> In the management cluster your App CRs are stored in a namespace with the same

Reviewer 3: The namespace is named after the cluster id right?

I am no native speaker, so take this with a grain of salt :salt:

```suggestion
App CRs are stored on the management cluster in the namespace named after the ID of your workload cluster.
```

Reviewer 3: oh ignore me, I should've read the whole sentence before commenting :facepalm:
